addons/product/report/product_pricelist_report.py

```diff
@@ -16,9 +20,83 @@ def get_html(self, data):
+        buffer = io.StringIO()
+        writer = csv.writer(buffer)
+
+        # Define headers dynamically based on the quantities
+        headers = ['Product', 'UOM'] + [f'Quantity ({qty} UOM)' for qty in data['quantities']]
+        writer.writerow(headers)
+
+        # Get the report data
+        report_data = self._get_report_data(data)
+        products = report_data['products']
+        quantities = report_data['quantities']
+
+        rows = []
+        for product in products:
+            variants = product.get('variants', [product])
+            for variant in variants:
+                row = [variant['name'], variant['uom']] + [variant['price'].get(qty, 0.0) for qty in quantities]
+                rows.append(row)
+
+        writer.writerows(rows)
+
+        # Get the CSV data from the buffer
+        csv_data = buffer.getvalue()
+        buffer.close()
+
+        file_name = f"Pricelist - {report_data['pricelist'].name}.csv"
+        return {'csv_data': csv_data, 'file_name': file_name}
+
+    @api.model
+    def export_xls(self, data):
+        buffer = io.BytesIO()
+        workbook = xlsxwriter.Workbook(buffer, {'in_memory': True})
+        worksheet = workbook.add_worksheet()
+
+        headers = ['Product', 'UOM'] + [f'Quantity ({qty} UOM)' for qty in data['quantities']]
+        worksheet.write_row(0, 0, headers)
+
+        # Initialize column widths based on header length
+        column_widths = [len(header) for header in headers]
+
+        # Gather data for setting dynamic widths
+        report_data = self._get_report_data(data)
+        products = report_data['products']
+        quantities = report_data['quantities']
+
+        # Collect all rows of data
+        rows = []
+        for product in products:
+            variants = product.get('variants', [product])
+            for variant in variants:
+                row = [variant['name'], variant['uom']] + [str(variant['price'].get(qty, 0.0)) for qty in quantities]
+                rows.append(row)
+
+                # Update column widths based on data length
+                for col_idx, cell_value in enumerate(row):
+                    column_widths[col_idx] = max(column_widths[col_idx], len(cell_value))
+
+        # Write the data rows to the worksheet
+        for row_idx, row_data in enumerate(rows, start=1):
+            worksheet.write_row(row_idx, 0, row_data)
+
+        # Set the column widths dynamically
+        for col_idx, width in enumerate(column_widths):
+            worksheet.set_column(col_idx, col_idx, width)
+
+        workbook.close()
+        xls_data = buffer.getvalue()
+        buffer.close()
+
+        file_name = f"Pricelist - {report_data['pricelist'].name}.xlsx"
+        return {'xls_data': base64.b64encode(xls_data).decode('utf-8'), 'file_name': file_name}
+
     def _get_report_data(self, data, report_type='html'):
         quantities = data.get('quantities', [1])
-
+        noProducts = data.get('noProducts', False)
```

Comment:
Shouldn't we hide the print buttons if there are no selected products ?
I don't think it makes sense to print an empty page.

Reply:
Yes, it doesn't make sense to print an empty page, but we need the `no_products` when the user enters the report preview screen from the `Print` button located in the header of the form view of the `product.pricelist`. At that time, no products would be included in the report, and based on `noProducts` we will be displaying a dialog box to add products to the report.